src/providers/pkcs11_provider/utils.rs

```diff
@@ -1,14 +1,173 @@
+    match alg {
+        #[allow(deprecated)]
+        Hash::Sha1 => Ok(CKG_MGF1_SHA1),
+        Hash::Sha256 => Ok(CKG_MGF1_SHA256),
+        Hash::Sha384 => Ok(CKG_MGF1_SHA384),
+        Hash::Sha512 => Ok(CKG_MGF1_SHA512),
+        _ => Err(ResponseStatus::PsaErrorNotSupported),
+    }
+}
+
+impl TryFrom<Algorithm> for CkMechanism {
+    type Error = ResponseStatus;
+
+    fn try_from(alg: Algorithm) -> Result<Self> {
+        match alg {
+            Algorithm::AsymmetricSignature(AsymmetricSignature::RsaPkcs1v15Sign { .. })
+            | Algorithm::AsymmetricEncryption(AsymmetricEncryption::RsaPkcs1v15Crypt { .. }) => {
+                Ok(CkMechanism::CkmRsaPkcs)
+            }
+            Algorithm::AsymmetricSignature(AsymmetricSignature::RsaPss {
+                hash_alg: SignHash::Specific(hash_alg),
+            }) => Ok(CkMechanism::CkmRsaPkcsPss(CK_RSA_PKCS_PSS_PARAMS {
+                hashAlg: CkMechanism::try_from(hash_alg)?.mech_type().into(),
+                mgf: hash_to_mgf1(hash_alg)?,
+                sLen: hash_alg.hash_length(),
+            })),
+            Algorithm::AsymmetricEncryption(AsymmetricEncryption::RsaOaep { hash_alg }) => {
+                Ok(CkMechanism::CkmRsaPkcsOaep(CK_RSA_PKCS_OAEP_PARAMS {
+                    hashAlg: CkMechanism::try_from(hash_alg)?.mech_type().into(),
+                    mgf: CKG_MGF1_SHA512,
+                    source: CKZ_DATA_SPECIFIED,
+                    pSourceData: std::ptr::null_mut(),
+                    ulSourceDataLen: 0,
+                }))
+            }
+            _ => Err(ResponseStatus::PsaErrorNotSupported),
+        }
+    }
+}
+
+impl From<CkMechanism> for CK_MECHANISM {
+    fn from(mech: CkMechanism) -> Self {
+        match mech {
+            CkMechanism::CkmRsaPkcs
+            | CkMechanism::CkmSha1
+            | CkMechanism::CkmSha256
+            | CkMechanism::CkmSha384
+            | CkMechanism::CkmSha512 => CK_MECHANISM {
+                mechanism: mech.mech_type().into(),
+                pParameter: std::ptr::null_mut(),
+                ulParameterLen: 0,
+            },
+            CkMechanism::CkmRsaPkcsPss(mut params) => {
+                let len = ::std::mem::size_of_val(&params);
+                let params: CK_RSA_PKCS_PSS_PARAMS_PTR = &mut params;
+                CK_MECHANISM {
+                    mechanism: mech.mech_type().into(),
+                    pParameter: params as CK_VOID_PTR,
+                    ulParameterLen: len,
+                }
+            }
+            CkMechanism::CkmRsaPkcsOaep(mut params) => {
+                let len = ::std::mem::size_of_val(&params);
+                let params: CK_RSA_PKCS_OAEP_PARAMS_PTR = &mut params;
+                CK_MECHANISM {
+                    mechanism: mech.mech_type().into(),
+                    pParameter: params as CK_VOID_PTR,
+                    ulParameterLen: len,
+                }
+            }
+        }
+    }
+}
```

Comment:
This works? More precisely, keeping a reference to those `params` structures, but dropping the object afterwards

Reply:
Hmm yes you are right that is really bad and unsafe ☠️ By the time the `params` are dereferenced, the address is no longer valid.
I think the abstraction `CkMechanism` is wrong here: it should really contain a reference on the parameters (with a lifetime specifier) and probably not be `Copy`.
The problem also is that `CK_MECHANISM` or `CK_ATTRIBUTE` do not contain a lifetime as they use raw pointers so I think it's hard to use the compiler to make sure the references are valid... In that case, even if you convert `CkMechanism` to a `CK_MECHANISM`, you lose the reference lifetime and the compiler can not do checks on it when accessing it

Reply:
Yeah, this is a massive issue, since you'd never be sure whether the reference is valid. I'm actually kinda shocked Rust lets you do this (the conversion to pointer part) without an `unsafe` around it... Maybe we should enable [this](https://rust-lang.github.io/rust-clippy/master/index.html#as_conversions) clippy lint? Though that might be overkill.